Docker/Image/Builder/ImageBuilder.php

```diff
@@ -396,50 +408,79 @@ protected function pullImage()
+                are ignored. */
+            $err = $process->getErrorOutput() . $process->getOutput();
+            if (preg_match('#KBC::USER_ERR:(.*?)KBC::USER_ERR#', $err, $matches)) {
+                $message = $matches[1];
+                throw new BuildParameterException($message);
+            } elseif ((strpos($err, WeirdException::ERROR_DEV_MAPPER) !== false) ||
+                    (strpos($err, WeirdException::ERROR_DEV_MAPPER_BUILD) !== false) ||
+                    (strpos($err, WeirdException::ERROR_DEVICE_RESUME) !== false)) {
+                // in case of this weird docker error, throw a new exception to retry the container
+                throw new WeirdException($err);
+            } else {
+                $message = "Build failed (code: {$process->getExitCode()}): " .
+                    " {$process->getOutput()} / {$process->getErrorOutput()}";
+                throw new BuildException($message);
+            }
+        }
+    }
+
     /*
      * @inheritdoc
      */
     public function prepare(array $configData)
     {
-        try {
-            $this->initParameters($configData);
-            parent::prepare($configData);
-            $temp = new Temp('docker');
-            $temp->initRunFolder();
-            $workingFolder = $temp->getTmpFolder();
-            $this->createDockerFile($workingFolder);
-
-            if (!$this->getCache()) {
-                $noCache = ' --no-cache';
-            } else {
-                $noCache = '';
-            }
-            $this->logger->debug("Building image");
-            $process = new Process(
-                "sudo docker build$noCache --tag=" . escapeshellarg($this->getFullImageId()) . " " . $workingFolder
-            );
-            // set some timeout to make sure that the parent image can be downloaded and Dockerfile can be built
-            $process->setTimeout(3600);
-            $process->run();
-            if ($process->getExitCode() != 0) {
-                /* string matching is used because currently it is not possible to have different exit codes for
-                    `docker build` It is either 0 for success or 1 for failure and the individual command exit codes
-                    are ignored. */
-                $err = $process->getErrorOutput();
-                if (preg_match('#KBC::USER_ERR:(.*?)KBC::USER_ERR#', $err, $matches)) {
-                    $message = $matches[1];
-                    throw new BuildParameterException($message);
-                } else {
-                    $message = "Build failed (code: {$process->getExitCode()}): " .
-                        " {$process->getOutput()} / {$process->getErrorOutput()}";
-                    throw new BuildException($message);
+        $this->initParameters($configData);
+        parent::prepare($configData);
+        $this->temp->initRunFolder();
+        $this->createDockerFile();
+        $retries = 0;
```

Comment:
To byl záznam myšlenkovýho pochodu! :-) Nechal jsem to tam jako vzkaz budoucím generacím.